graphbolt/src/shared_memory_utils.cc

```diff
@@ -0,0 +1,178 @@
+                size * torch::elementSize(std::get<2>(meta));
+    } else {
+      optional_tensors.push_back(torch::nullopt);
+    }
+  }
+  return std::make_pair(std::move(shm), std::move(optional_tensors));
+}
+
+SharedMemoryTensors CopyTensorsToSharedMemory(
+    const std::string& name,
+    const std::vector<torch::optional<torch::Tensor>>& tensors,
+    int64_t max_meta_memory_size) {
+  torch::serialize::OutputArchive archive;
+  archive.write("num_tensors", static_cast<int64_t>(tensors.size()));
+  for (size_t i = 0; i < tensors.size(); ++i) {
+    archive.write(
+        "tensor_" + std::to_string(i) + "_has_value", tensors[i].has_value());
+    if (tensors[i].has_value()) {
+      archive.write(
+          "tensor_" + std::to_string(i) + "_shape", tensors[i].value().sizes());
+      archive.write(
+          "tensor_" + std::to_string(i) + "_dtype",
+          tensors[i].value().scalar_type());
+    }
+  }
+  auto meta_shm = CopyTorchArchiveToSharedMemory(
+      name + "_meta", max_meta_memory_size, archive);
+  auto data_shm = CopyTensorsDataToSharedMemory(name + "_data", tensors);
+
+  std::vector<torch::optional<torch::Tensor>> ret_tensors;
+  auto mem_buf = data_shm->GetMemory();
+  for (auto optional_tensor : tensors) {
+    if (optional_tensor.has_value()) {
+      auto tensor = optional_tensor.value();
+      ret_tensors.push_back(
+          torch::from_blob(mem_buf, tensor.sizes(), tensor.dtype()));
+      int64_t size = tensor.numel() * tensor.element_size();
+      mem_buf = static_cast<char*>(mem_buf) + size;
+    } else {
+      ret_tensors.push_back(torch::nullopt);
+    }
+  }
+  return std::make_tuple(
+      std::move(meta_shm), std::move(data_shm), std::move(ret_tensors));
+}
+
+SharedMemoryTensors LoadTensorsFromSharedMemory(
+    const std::string& name, int64_t meta_memory_size) {
+  torch::serialize::InputArchive archive;
+  auto meta_shm = LoadTorchArchiveFromSharedMemory(
+      name + "_meta", meta_memory_size, archive);
+  std::vector<std::tuple<bool, std::vector<int64_t>, torch::ScalarType>> metas;
+  int64_t num_tensors = read_from_archive(archive, "num_tensors").toInt();
+  for (int64_t i = 0; i < num_tensors; ++i) {
+    bool has_value =
+        read_from_archive(archive, "tensor_" + std::to_string(i) + "_has_value")
+            .toBool();
+    if (has_value) {
+      auto shape =
+          read_from_archive(archive, "tensor_" + std::to_string(i) + "_shape")
+              .toIntVector();
+      auto dtype =
+          read_from_archive(archive, "tensor_" + std::to_string(i) + "_dtype")
+              .toScalarType();
+      metas.push_back({true, shape, dtype});
+    } else {
+      metas.push_back({false, {}, torch::ScalarType::Undefined});
+    }
+  }
+  SharedMemoryPtr data_shm;
+  std::vector<torch::optional<torch::Tensor>> ret_tensors;
+  std::tie(data_shm, ret_tensors) =
+      LoadTensorsDataFromSharedMemory(name + "_data", metas);
+  return std::make_tuple(
```

Comment:
I am wondering if we could utilize `std::tuple_cat()` here. or we could avoid explicit definition of `data_shm` and `ret_tensors` which looks a bit unnecessary via `auto&& tensor_data = ...` and access via `std::pair<>.first/second`?

Reply:
I just learned `std::tuple_cat()`. It is a great function to simplify code in this situation, which involves tuple or pair as returned value. We may optimize this code in the future.